lib/docker-delta.coffee

```diff
@@ -119,27 +119,37 @@ hardlinkCopy = (srcRoot, dstRoot, linkDests) ->
 	rsync.waitAsync()
 
 applyBatch = (rsync, batch, timeout) ->
-	new Promise (resolve) ->
-		batch.pipe(rsync.stdin).on('unpipe', resolve)
-	.then ->
-		if timeout is 0
-			# wait for clean exit
-			return rsync.waitAsync()
-		else
-			# or until the given timeout is exceeded
-			return rsync.waitAsync().timeout(timeout)
-	.catch Promise.TimeoutError, ->
-		# timed out; kill it.
+	p = new Promise (resolve, reject) ->
+		batch.pipe(rsync.stdin)
+		.on('error', reject)
+		.on('finish', resolve)
+
+	if timeout isnt 0
+		p = p.timeout(timeout)
+
+	return p.then ->
+		# `p` is resolved on 'finish' but that doesn't guarantee rsync exited cleanly :/
+		# normally we'd end up here on rsync exiting with 0, but there are cases
+		# our pipe to its stdin is broken with no error and rsync still is up.
+		# so to find the reason rsync exited, we "poll" it once here via `waitAsync`.
+		# if rsync exited cleanly, `waitAsync` will resolve and the chain continues.
+		# for any other case, a short timeout will ensure we don't wait forever and
+		# manually kill rsync below.
+		rsync.waitAsync().timeout(5000)
+	.catch (err) ->
```

Comment:
This will catch the case where `waitAsync` rejects due to a non-0 exit code, is that intentional?

Reply:
Yes, the error will be rethrown after rsync exits a couple of lines below, unless rsync's exit itself causes an error.

Reply:
The exit below will surely cause an error though, because the process would already have exited and trying to exit an exited program seems like it would cause an error?

Reply:
Not sure I understand -- you mean you suspect `rsync.kill` below will throw? From [node docs](https://nodejs.org/api/child_process.html#child_process_subprocess_kill_signal):

> Sending a signal to a child process that has already exited is not an error

The caveat is that the PID the signal is sent to might have been reassigned to a new process, but I highly doubt that's a problem in practice given I haven't ever seen an OS that doesn't increment PIDs monotonically.

Reply:
Ok, I didn't expect that, as for pid's incrementing monotonically that is true, however they can wrap around and reuse earlier pids - I'm not worried about that in reality though